### PR TITLE
Support running on M1 (via Rosetta) with native lima and qemu

### LINKED ISSUE #957: Package M1 build of Rancher Desktop
The required steps are shown in the description of #933:

* Start with the `x86_64` build
* Replace ISO and lima directory with `aarch64` versions
* Use arm64 versions of `helm`, `kuberlr`, etc
* Create DMG

For #831 we'll need some mechanism so that RD can tell if the assets are `x86_64` or `aarch64`.

### PR DESCRIPTION
**Pull request by jandubois**
The changes in this PR let us run the `x86_64` build of `Rancher Desktop.app` on `aarch64` after replacing the ISO and the lima directory:

```console
$ cd "/Applications/Rancher Desktop.app/Contents/Resources/resources/darwin"
$ mv alpine-lima-v0.2.1-rd-3.13.5.iso alpine-lima-v0.2.1-rd-3.13.5.iso.orig
$ cp ~/Downloads/alpine-lima-rd-3.13.5-aarch64.iso alpine-lima-v0.2.1-rd-3.13.5.iso
$ mv lima lima.orig
$ mkdir lima
$ cd lima
$ tar xvfz ~/Downloads/lima-and-qemu.macos.aarch64.tar.gz
```
Download links:
https://github.com/lima-vm/alpine-lima/releases/download/v0.2.1/alpine-lima-rd-3.13.5-aarch64.iso
https://github.com/rancher-sandbox/lima-and-qemu/releases/download/v1.12/lima-and-qemu.macos-aarch64.tar.gz

Closes #957

### CODE REVIEW

**Comment by DevPre24** [comment]
@jandubois Hi, thanks for your efforts porting it to the m1.

I was doing a little bit of testing, but unfortunately on my system (MBP 13 M1) the limavm virtual machine doesn't seem to start, i get this error:

```[Bds]Booting UEFI Misc Device
 BlockSize : 262144 
 LastBlock : FF 
[Bds] Expand VenHw(93E34C7E-B50E-11DF-9223-2443DFD72085,00) -> <null string>
BdsDxe: failed to load Boot0001 "UEFI Misc Device" from VenHw(93E34C7E-B50E-11DF-9223-2443DFD72085,00): Not Found
Memory  Previous  Current    Next   
 Type    Pages     Pages     Pages  
======  ========  ========  ========
  09    00000000  00000090  000000B4
  0A    00000000  00000030  0000003C
  00    00000000  00000300  000003C0
  06    0000012C  00000330  000003FC
  05    00000096  00000250  000002E4
  03    000003E8  00000483  000005A3
  04    00002EE0  0000217D  00002EE0
  01    00000014  00000000  00000014
  02    00000000  00000000  00000000
Memory Type Information settings change.
[Bds]Booting UEFI Misc Device 2
 BlockSize : 512 
 LastBlock : 53FFF 
PartitionDxe: El Torito standard found on handle 0x13E2D6E98.
 BlockSize : 512 
 LastBlock : F 
 BlockSize : 512 
 LastBlock : 2CFF 
FatDiskIo: Cache Page OutBound occurred! 
 BlockSize : 512 
 LastBlock : 53FFF 
PartitionDxe: El Torito standard found on handle 0x13E2D6E98.
 BlockSize : 512 
 LastBlock : F 
 BlockSize : 512 
 LastBlock : 2CFF 
[Bds] Expand PciRoot(0x0)/Pci(0x4,0x0) -> <null string>
BdsDxe: failed to load Boot0002 "UEFI Misc Device 2" from PciRoot(0x0)/Pci(0x4,0x0): Not Found
Memory  Previous  Current    Next   
 Type    Pages     Pages     Pages  
======  ========  ========  ========
  09    00000000  00000090  000000B4
  0A    00000000  00000030  0000003C
  00    00000000  00000300  000003C0
  06    0000012C  00000330  000003FC
  05    00000096  00000250  000002E4
  03    000003E8  00000483  000005A3
  04    00002EE0  0000217D  00002EE0
  01    00000014  00000000  00000014
  02    00000000  00000000  00000000
Memory Type Information settings change.```

Workflow:
- Cloned the project
- Checked out the `m1 branch`
- Built using nodejs 14 (both amd64 and arm64, btw the electron component seems to run natively as well, no clue if it has some issues so i tried both)
- Replaced `alpine-lima-rd` and `lima-and-qemu` in the Applications files with those provided

**Comment by bdhoine** [comment]
I managed to take this branch as a base to run Rancher Desktop as a native app on my m1. I did patch some stuff where the electron helper methods are used and some other patches to have the binaries be auto downloaded by the install scripts. The vm starts and I was able to run an nginx pod and expose it on port 8080.

File: src/k8s-engine/lima.ts [hunk 1]
```diff
@@ -713,12 +715,13 @@ export default class LimaBackend extends events.EventEmitter implements K8s.Kube
 
     try {
       const scriptPath = path.join(workdir, 'install-k3s');
+      const k3s = Electron.app.runningUnderRosettaTranslation ? 'k3s-arm64' : 'k3s';
```
Comment (mook-as) [inline src/k8s-engine/lima.ts:718]: This should depend on the `arch` we passed in the constructor instead.

File: src/k8s-engine/lima.ts [hunk 2]
```diff
@@ -421,7 +423,7 @@ export default class LimaBackend extends events.EventEmitter implements K8s.Kube
     const config: LimaConfiguration = merge({}, baseConfig, DEFAULT_CONFIG as LimaConfiguration, {
       images: [{
         location: this.baseDiskImage,
-        arch:     'x86_64',
+        arch:     Electron.app.runningUnderRosettaTranslation ? 'aarch64' : 'x86_64',
```
Comment (mook-as) [inline src/k8s-engine/lima.ts:426]: This should depend on the `arch` we passed in the constructor instead.

**Review by mook-as — CHANGES_REQUESTED** [review]

File: src/assets/scripts/install-k3s [hunk 3]
```diff
@@ -20,12 +20,19 @@ fi
 mkdir -p /etc/rancher/k3s
 rm -f /etc/rancher/k3s/k3s.yaml
 
+K3S=k3s
+ARCH=amd64
+if [ "$(uname -m)" = "aarch64" ]; then
```
[the remark this answers is not in the document]
Comment (jandubois) [reply src/assets/scripts/install-k3s:25]: We'll have to change things. Just like we'll have to change the code again once we build a native GUI.

File: src/k8s-engine/lima.ts [hunk 4]
```diff
@@ -11,6 +11,7 @@ import timers from 'timers';
 import util from 'util';
 import { ChildProcess, spawn as spawnWithSignal } from 'child_process';
 
+import Electron from 'electron';
```
Comment (mook-as) [inline src/k8s-engine/lima.ts:14]: Is this import still used?

**Review by mook-as — CHANGES_REQUESTED** [review]

**Comment by jandubois** [comment]
Converted to "draft" mode because I will add packaging changes here as well.

File: src/k8s-engine/lima.ts (hunk 4 above)
Comment (jandubois) [reply src/k8s-engine/lima.ts]: No (removed)

**Review by mook-as — DISMISSED** [review]
Tried the build from CI, and it worked (after manually skipping the prompt due to unverified developer).

It doesn't have correct permissions to handle the integrations (`/usr/local/bin` stuff).  But that can be dealt with later.

File: .github/workflows/package.yaml [hunk 5]
```diff
@@ -47,12 +48,26 @@ jobs:
...
+      if: startsWith(matrix.os, 'macos-11')
+      run: M1=1 npm ci
     - name: npm run build
       run: |
         case "${{ matrix.os }}" in
-          macos-*)
+          macos-10*)
+            # Build Intel version with Intel assets
             npm run build -- --mac --publish=never
+            mv dist/Rancher*.zip "dist/$(basename dist/Rancher*.zip .zip)-x86_64.zip"
```
[the remark this answers is not in the document]
Comment (jandubois) [reply .github/workflows/package.yaml:63]: > if we're going to ignore the file name when we do the upload

Not sure what you mean: we do upload with a wildcard, but this is still the filename that will be used in the assets list (and in the checksums file).

[the remark this answers is not in the document]
Comment (jandubois) [reply .github/workflows/package.yaml:63]: Without this change the filenames are identical for both builds.

[the remark this answers is not in the document]
Comment (jandubois) [reply .github/workflows/package.yaml]: No, we don't need the renames; they do nothing for us. I've removed them.

**Comment by jandubois** [comment]
> It doesn't have correct permissions to handle the integrations (/usr/local/bin stuff). But that can be dealt with later.

I think you will observe the same issues when you install on a fresh install on Intel as well (pristine Big Sur VM):

```console
% sw_vers 
ProductName:	macOS
ProductVersion:	11.3
BuildVersion:	20E232
% ls -la /usr/local
total 0
drwxr-xr-x   2 root  wheel   64  1 Jan  2020 .
drwxr-xr-x@ 11 root  wheel  352  1 Jan  2020 ..
```

**Comment by jandubois** [comment]
>the limavm virtual machine doesn't seem to start, i get this error (serial console)

@DevPre24 It didn't work because you were building natively on M1, but the code was only checking if it was running under Rosetta. I've updated the code today to also check `os.arch()`, so maybe it works better now.

However, I did try to cross-build a native app on Intel, and that didn't work. I gave up on it because we also are using a few node modules that are platform-dependent, so those wouldn't work unless compiled on ARM.

For now we will ship the GUI running via Rosetta because we have no way to run Github Actions on M1 machines, so couldn't produce builds using CI.

**Review by mook-as — APPROVED** [review]
Looks reasonable to me; we can do follow-ups as needed.

**Pull request merged**